bionic/deriver.py

```diff
@@ -150,52 +146,51 @@ def _get_ready_for_bootstrap_resolution(self):
         if self._is_ready_for_bootstrap_resolution:
             return
 
-        # Generate the static key spaces and tasks for each entity.
+        # Generate the static key spaces and tasks for each descriptor.
         self._key_spaces_by_dnode = {}
         self._task_lists_by_dnode = {}
         for name in self._flow_state.providers_by_name.keys():
             dnode = DescriptorNode.from_descriptor(name)
             self._populate_dnode_info(dnode)
 
-        # Create a state object for each task.
-        self._task_states_by_key = {}
+        # Create a lookup table for all tasks.
+        self._tasks_by_key = {}
         for tasks in self._task_lists_by_dnode.values():
             for task in tasks:
-                task_state = TaskState(task)
-                for key in task.keys:
-                    self._task_states_by_key[key] = task_state
+                for task_key in task.keys:
+                    self._tasks_by_key[task_key] = task
 
-        # Initialize each of the task states.
-        for tasks in self._task_lists_by_dnode.values():
-            for task in tasks:
-                task_state = self._task_states_by_key[task.keys[0]]
-                self._initialize_task_state(task_state)
+        # Create a state object for each task.
+        self._task_states_by_key = {}
+        for task_key in self._tasks_by_key.keys():
+            self._get_or_create_task_state_for_key(task_key)
 
         self._is_ready_for_bootstrap_resolution = True
 
-    def _initialize_task_state(self, task_state):
-        if task_state.is_initialized:
-            return
-
-        task = task_state.task
-
-        dep_states = [self._task_states_by_key[dep_key] for dep_key in task.dep_keys]
-
-        for dep_state in dep_states:
-            self._initialize_task_state(dep_state)
-
-            task_state.parents.append(dep_state)
-            dep_state.children.append(task_state)
+    def _get_or_create_task_state_for_key(self, task_key):
+        task_state = self._task_states_by_key.get(task_key, None)
+        if task_state is not None:
+            return task_state
 
+        task = self._tasks_by_key[task_key]
+        dep_states = [
+            self._get_or_create_task_state_for_key(dep_key) for dep_key in task.dep_keys
+        ]
         # All names in this task should point to the same provider.
-        (task_state.provider,) = set(
+        (provider,) = set(
             self._flow_state.get_provider(task_key.dnode.to_entity_name())
             for task_key in task.keys
         )
         # And all the task keys should have the same case key.
-        (task_state.case_key,) = set(task_key.case_key for task_key in task.keys)
+        (case_key,) = set(task_key.case_key for task_key in task.keys)
+
+        task_state = TaskState(
+            task=task, dep_states=dep_states, provider=provider, case_key=case_key,
```

Comment:
Nope, Python won't match the names for you. If you write

    task_state = TaskState(task, dep_states, provider, case_key)

then Python doesn't care what the variable names are, just their ordering -- which in this case happens to be wrong. I added the redundant names so that I can't screw it without being obvious.

If Bionic were a language it would totally do this for you, though.